[Tune][Air] Fix MLflowLoggerCallback to enable its use with PBT (#27783)

sjp611 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->
In the current MLflowLoggerCallback implementation, the function to store parameters in MLflow (mlflow_util.log_param) is only used in log_trial_start.

The problem is that PBT explores various parameters. The initially stored parameters are almost always different from the optimal parameters.

We need to be able to verify not only the initial parameters but also the optimal parameters in MLflow. The current implementation only allows us to verify the initial parameters.

This contribution modifies the MLflowLoggerCallback function to enable the storage of both the final and initial parameters in MLflow.

## Related issue number
"Open #27783"

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

sjp611 committed:
[Tune][Air] Fix MLflowLoggerCallback to enable its use with PBT (#27783)

Signed-off-by: Sungjoon Park <sjp611@gmail.com>

sjp611 committed:
Merge branch 'master' into new-feature

sjp611 committed:
Merge branch 'master' into new-feature

sjp611 committed:
Merge branch 'master' into new-feature

sjp611 committed:
Merge branch 'master' into new-feature

sjp611 commented:
In the current MLflowLoggerCallback implementation, the function to store parameters in MLflow (mlflow_util.log_param) is only used in log_trial_start.

The problem is this: PBT explores various parameters. The initially stored parameters are almost always different from the optimal parameters.

We need to be able to verify not only the initial parameters but also the optimal parameters in MLflow. The current implementation only allows us to verify the initial parameters.

This contribution modifies the MLflowLoggerCallback function to enable the storage of both the final and initial parameters in MLflow.

sjp611 commented:
I edited comment "Why are these changes needed?"

sjp611 committed:
Merge branch 'master' into new-feature

sjp611 committed:
Merge branch 'master' into new-feature

sjp611 committed:
Merge branch 'master' into new-feature

TimothySeah reviewed (commented):
Also please add a unit test for this, thanks!

sjp611 committed:
Merge branch 'master' into new-feature

sjp611 commented on python/ray/air/integrations/mlflow.py:
```diff
@@ -222,6 +222,8 @@ class MLflowLoggerCallback(LoggerCallback):
         save_artifact: If set to True, automatically save the entire
             contents of the Tune local_dir as an artifact to the
             corresponding run in MlFlow.
+        log_params_first: If set to False, the parameters of MLflow are logged
```
I have implemented your suggested changes.

sjp611 commented:
@TimothySeah  I have implemented your suggested changes and written unit tests. thanks!

sjp611 committed:
[Tune][Air] Fix MLflowLoggerCallback to enable its use with PBT (#27783)

Why are these changes needed?

In the current MLflowLoggerCallback implementation, the function to store parameters in MLflow (mlflow_util.log_param) is only used in log_trial_start.

The problem is that PBT explores various parameters. The initially stored parameters are almost always different from the optimal parameters.

We need to be able to verify not only the initial parameters but also the optimal parameters in MLflow. The current implementation only allows us to verify the initial parameters.

This contribution modifies the MLflowLoggerCallback function to enable the storage of both the final and initial parameters in MLflow.

Related issue number
"Open #27783"

Checks
  - [x] I've signed off every commit(by using the -s flag, i.e., git commit -s) in this PR.
  - [x]  I've run scripts/format.sh to lint the changes in this PR.
 - [  ] I've included any doc changes needed for https://docs.ray.io/en/master/.
 - [  ] I've added any new APIs to the API Reference. For example, if I added a method in Tune, I've added it in doc/source/tune/api/ under the corresponding .rst file.
 I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/

Testing Strategy
- [x] Unit tests
- [  ] Release tests
- [  ] This PR is not tested :(

Signed-off-by: Sung Joon Park <sjp611@gmail.com>

sjp611 committed:
[Tune][Air] Fix MLflowLoggerCallback to enable its use with PBT (#27783)

Why are these changes needed?

In the current MLflowLoggerCallback implementation, the function to store parameters in MLflow (mlflow_util.log_param) is only used in log_trial_start.

The problem is that PBT explores various parameters. The initially stored parameters are almost always different from the optimal parameters.

We need to be able to verify not only the initial parameters but also the optimal parameters in MLflow. The current implementation only allows us to verify the initial parameters.

This contribution modifies the MLflowLoggerCallback function to enable the storage of both the final and initial parameters in MLflow.

Related issue number
"Open ray-project#27783"

Checks
  - [x] I've signed off every commit(by using the -s flag, i.e., git commit -s) in this PR.
 - [x]  I've run scripts/format.sh to lint the changes in this PR.
 - [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
 - [ ] I've added any new APIs to the API Reference. For example, if I added a method in Tune, I've added it in doc/source/tune/api/ under the corresponding .rst file.
 I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/

Testing Strategy
- [x] Unit tests
- [ ] Release tests
- [ ] This PR is not tested :(

Signed-off-by: Sung Joon Park <sjp611@gmail.com>

sjp611 committed:
Merge branch 'master' into new-feature

sjp611 commented on python/ray/air/tests/test_integration_mlflow.py:
```diff
@@ -193,6 +193,42 @@ def testMlFlowLoggerLogging(self):
+        logger.setup()
+        exp_id = logger.mlflow_util.experiment_id
+
+        # Check if run is created with proper tags.
```
Thanks for the careful review! I've applied the changes as you suggested.
